### DISCUSSION TITLE
403 error

### DISCUSSION THREAD

**Discussion by GavinReid82** [community/community#127413] [New to GitHub] [answered]
Hi, I'm new to GitHub and programming in general. I'm working through a Codecademy vid (https://www.youtube.com/watch?v=eNw8MgguLsQ) but am unable to push a new branch to my repository, getting the following error:

'''
(base) gavinreid@MacBook-Air-3 gitPractice % git push --set-upstream origin print_changes
remote: Permission to GavinReid82/gitPractice.git denied to GavinReid82.
fatal: unable to access 'https://github.com/GavinReid82/gitPractice.git/': The requested URL returned error: 403
'''

I changed my password in keychain to a new token I just generated (following other discussions), but still can't get access.

Any help would be appreciated.
Gavin

**Comment by Sbiswas001** [comment, marked as answer]
The error message "Permission to GavinReid82/gitPractice.git denied to GavinReid82" typically indicates that there's an authentication problem.

Here are steps to troubleshoot and resolve this issue:

1. **Generate a Personal Access Token (PAT):**
   Ensure that you have generated a new personal access token from GitHub. Follow these steps if you haven't done this yet:
   - Go to your GitHub [Settings](https://github.com/settings/profile).
   - Navigate to **Developer settings** > **Personal access tokens**.
   - Click on **Generate new token**.
   - Select the necessary scopes, such as `repo` for full control of private repositories.
   - Copy the generated token.

2. **Update Git Credentials:**
   Use the token as your password when performing Git operations that require authentication.
   - Open your terminal and run the following command to set the token as your credential:
     ```sh
     git config --global credential.helper store
     ```
   - Then, try to push again. When prompted for a username, use your GitHub username. For the password, use the personal access token.

3. **Clear Previous Credentials:**
   If you previously stored credentials that might be incorrect, you can clear them and re-enter the correct credentials.
   - On macOS, you can clear stored GitHub credentials using the following command:
     ```sh
     git credential-osxkeychain erase
     ```
   - When prompted, enter the following:
     ```
     protocol=https
     host=github.com
     ```

4. **Use HTTPS with the Token:**
   Ensure that your GitHub repository URL is set to use HTTPS:
   - Check your remote URL:
     ```sh
     git remote -v
     ```
   - If it shows an SSH URL, you need to change it to HTTPS:
     ```sh
     git remote set-url origin https://github.com/GavinReid82/gitPractice.git
     ```

5. **Verify Token in Keychain:**
   Since you mentioned updating your keychain, double-check that the personal access token is correctly stored in your keychain and associated with GitHub.

6. **Re-clone the Repository:**
   If all else fails, sometimes re-cloning the repository with the new token can resolve any lingering issues:
   - Backup your changes.
   - Remove the current repository.
   - Clone the repository again using HTTPS and your new token:
     ```sh
     git clone https://github.com/GavinReid82/gitPractice.git
     ```

After following these steps, try to push your changes again:
```sh
git push --set-upstream origin print_changes
```

**Reply by GavinReid82** [reply]
That worked after the second step! Thank you so much for your helpful and detailed response, I really appreciate it!
